### PR TITLE
[Feature 6] Add/Edit Expenses

### LINKED ISSUE #7: [Feature 6] Add/edit Expenses
A feature to add/edit expenses.

### PR DESCRIPTION
**Pull request by luluwuwu**
UCI implementation complete. Still working on the last two UCITests for editExpenseinRecurring. Please JavaDoc warnings and errors, will rewrite them later.

### CODE REVIEW

**Review by YinLingithub — CHANGES_REQUESTED** [review]
I have revert your commits on main. It seems that you made commits to main where you should have made commits here. Make sure that you are on your branch in your local file and only make commits to your branch.

File: src/main/java/use_case/add_edit_expenses_use_case/ExpenseIB.java
Comment (katarinaavucic) [inline src/main/java/use_case/add_edit_expenses_use_case/ExpenseIB.java]: Instead of saying what it returns you should say what it does.

Comment (katarinaavucic) [reply src/main/java/use_case/add_edit_expenses_use_case/ExpenseIB.java]: same for editExpenseInMonth.

File: src/main/java/use_case/add_edit_expenses_use_case/ExpenseID.java
Comment (katarinaavucic) [inline src/main/java/use_case/add_edit_expenses_use_case/ExpenseID.java]: Please remove these lines as they are between the Javadoc and method.

File: src/main/java/use_case/add_edit_expenses_use_case/ExpenseID.java
Comment (katarinaavucic) [inline src/main/java/use_case/add_edit_expenses_use_case/ExpenseID.java]: Are we not putting Javadocs for getters and setters? I know I have them but I'm not sure about Yin and Ari. We should decide whether we are for consistency.

File: src/main/java/use_case/add_edit_expenses_use_case/ExpenseOD.java
Comment (katarinaavucic) [inline src/main/java/use_case/add_edit_expenses_use_case/ExpenseOD.java]: Missing constructor Javadoc.

Comment (katarinaavucic) [reply src/main/java/use_case/add_edit_expenses_use_case/ExpenseOD.java]: And getter Javadoc.

File: src/main/java/use_case/add_edit_expenses_use_case/ExpenseUCI.java
Comment (katarinaavucic) [inline src/main/java/use_case/add_edit_expenses_use_case/ExpenseUCI.java:1]: Missing class Javadoc. Line 10: missing "private" Line 30: helper methods should be private. Line 125: method missing @throws param. Line 189: missing Javadoc description and @return and @throws commented out. Line 202: missing Javadoc description. Line 189-211: these methods should be private. Overall the overridden methods are very lengthy. I think you should create private helper methods for dealing with the recurring expense part.

File: src/main/java/views/add_edit_epense_views/ExpenseP.java
Comment (katarinaavucic) [inline src/main/java/views/add_edit_epense_views/ExpenseP.java:1]: typo.

File: src/main/java/views/add_edit_epense_views/ExpenseP.java
Comment (katarinaavucic) [inline src/main/java/views/add_edit_epense_views/ExpenseP.java:11]: Missing class Javadoc.

File: src/main/java/views/add_edit_epense_views/AddExpenseV.java
Comment (katarinaavucic) [inline src/main/java/views/add_edit_epense_views/AddExpenseV.java:1]: Line 13: Missing class Javadoc. Line 25: Missing constructor Javadoc (why is it under the method declaration?). Line 75: why is your Javadoc under the method ?? missing @param too.

File: src/main/java/views/add_edit_epense_views/EditExpenseV.java
Comment (katarinaavucic) [inline src/main/java/views/add_edit_epense_views/EditExpenseV.java:1]: Missing class and constructor Javadocs (place them above the method). Line 78: Javadoc is under the method and missing @param too. Line 84: are you supposed to be using if or elif here? If its several if statements they shouldn't be nested like that because it looks like they should be else if statements. Also, the else statement is only attached to the last if.

File: src/main/java/views/add_edit_epense_views/ExpenseC.java
Comment (katarinaavucic) [inline src/main/java/views/add_edit_epense_views/ExpenseC.java:1]: Line 10: Javadoc is under the class. Also you used pothole case. Line 14: Should be private. Line 25: it sounds like this is only for edit expense? Maybe fix Javadoc to mention add and edit. Also, the method name "expenseInMonth" doesn't really tell you that it is adding or editing the expense, so I would make that more descriptive.

**Review by katarinaavucic — CHANGES_REQUESTED** [review]
It was a little hard to review bc i couldn't write stuff on each line, so let me know if you have trouble understanding which lines i mean. I took note of where you were missing Javadocs to help you finish them up (and make sure they were all completed). You also put them in the wrong spot quite frequently so please go over that. Please take a look at my comments. Good job you are in the final stretch!

File: src/main/java/use_case/add_edit_expenses_use_case/ExpenseID.java
Comment (aricasas) [reply src/main/java/use_case/add_edit_expenses_use_case/ExpenseID.java]: I think we are

File: src/main/java/views/add_edit_epense_views/ExpenseP.java
Comment (katarinaavucic) [reply src/main/java/views/add_edit_epense_views/ExpenseP.java:1]: As in your whole folder has a typo so you will need to rename it.

File: src/main/java/views/add_edit_epense_views/AddExpenseV.java
Comment (aricasas) [inline src/main/java/views/add_edit_epense_views/AddExpenseV.java:1]: Instance attributes missing private/private final

Comment (aricasas) [inline src/main/java/views/add_edit_epense_views/AddExpenseV.java:1]: Instance attributes missing private/private final

Comment (aricasas) [inline src/main/java/views/add_edit_epense_views/AddExpenseV.java:1]: Inside the actionPerformed method, instead of isRecurring == (!Objects.equals(selectedCategory, "Other"))
you could do isRecurring != Objects.equals(selectedCategory, "Other")

File: src/main/java/views/add_edit_epense_views/EditExpenseV.java
Comment (aricasas) [inline src/main/java/views/add_edit_epense_views/EditExpenseV.java:1]: Instance attributes are missing private/private final

Comment (aricasas) [inline src/main/java/views/add_edit_epense_views/EditExpenseV.java:1]: On this one too, inside the actionPerformed method, instead of isRecurring == (!Objects.equals(selectedCategory, "Other"))
you could do isRecurring != Objects.equals(selectedCategory, "Other")

Comment (aricasas) [inline src/main/java/views/add_edit_epense_views/EditExpenseV.java:1]: Inside the actionPerformed in the else, you could do it just like in AddExpenseV so that you don't need to set message to null and do an extra if outside the try catch

File: src/main/java/use_case/add_edit_expenses_use_case/ExpenseOD.java
Comment (aricasas) [inline src/main/java/use_case/add_edit_expenses_use_case/ExpenseOD.java]: Missing private final

File: src/main/java/use_case/add_edit_expenses_use_case/ExpenseID.java
Comment (luluwuwu) [reply src/main/java/use_case/add_edit_expenses_use_case/ExpenseID.java]: okay that means I'll need to add them in add/edit category as well

File: src/test/java/use_case/add_edit_expenses_use_case/ExpenseUCITest.java
Comment (aricasas) [inline src/test/java/use_case/add_edit_expenses_use_case/ExpenseUCITest.java:1]: In line 64, the comment is wrong. The expected value is 1 and that explanation isn't needed

Comment (aricasas) [reply src/test/java/use_case/add_edit_expenses_use_case/ExpenseUCITest.java:1]: Same in line 83

Comment (aricasas) [inline src/test/java/use_case/add_edit_expenses_use_case/ExpenseUCITest.java:1]: addExpenseInMonthInvalidDoubleFail has the wrong JavaDoc

Comment (aricasas) [reply src/test/java/use_case/add_edit_expenses_use_case/ExpenseUCITest.java:1]: Same in line 101

Comment (aricasas) [reply src/test/java/use_case/add_edit_expenses_use_case/ExpenseUCITest.java:1]: The one in line 138 has the right number, but the explanation is wrong

Comment (aricasas) [inline src/test/java/use_case/add_edit_expenses_use_case/ExpenseUCITest.java:1]: There's a bunch of commented lines from 108-121

Comment (aricasas) [inline src/test/java/use_case/add_edit_expenses_use_case/ExpenseUCITest.java:1]: Shouldn't have pothole case "expense_name" in line 143

Comment (aricasas) [reply src/test/java/use_case/add_edit_expenses_use_case/ExpenseUCITest.java:1]: Also at the very end of the file there's more

**Review by aricasas — CHANGES_REQUESTED** [review]
Most of my comments were about JavaDocs, but functionality seems to be done!

File: src/main/java/views/add_edit_epense_views/EditExpenseV.java
Comment (katarinaavucic) [reply src/main/java/views/add_edit_epense_views/EditExpenseV.java:1]: Ignore the part about the if statements, I just saw the EditCategoryV so i understand it now. I think it is a little confusing how the ifs are stacked like else ifs, but the functionality of it is fine.

Comment (YinLingithub) [reply src/main/java/views/add_edit_epense_views/EditExpenseV.java:1]: @luluwuwu I don't think the recurring expenses must be saved in "Other" Category, so this if statement can be removed.

File: src/main/java/use_case/add_edit_expenses_use_case/ExpenseUCI.java
Comment (YinLingithub) [inline src/main/java/use_case/add_edit_expenses_use_case/ExpenseUCI.java:1]: line 76:
 _if(recurringExpenseList.size() > 0) {_
This if statement is not needed.

Comment (luluwuwu) [reply src/main/java/use_case/add_edit_expenses_use_case/ExpenseUCI.java:1]: Line 30: helper methods should be private. -- They're used in tests

File: src/main/java/use_cases/add_edit_expenses_use_case/ExpenseIB.java
Comment (katarinaavucic) [inline src/main/java/use_cases/add_edit_expenses_use_case/ExpenseIB.java:12]: Sorry for being so nitpicky with this but maybe you could write "Adds an expense to the current month and returns a response describing if it was added successfully or not."

Comment (katarinaavucic) [reply src/main/java/use_cases/add_edit_expenses_use_case/ExpenseIB.java:12]: And same goes for the edit version below.

File: src/main/java/use_cases/add_edit_expenses_use_case/ExpenseID.java
Comment (katarinaavucic) [inline src/main/java/use_cases/add_edit_expenses_use_case/ExpenseID.java:55]: Do you need this setter?

File: src/main/java/use_cases/add_edit_expenses_use_case/ExpenseUCI.java
Comment (katarinaavucic) [inline src/main/java/use_cases/add_edit_expenses_use_case/ExpenseUCI.java:254]: Yay nice private helper method!

**Review by katarinaavucic — COMMENTED** [review]
Looks good so far, I am having a hard time reading your code atm so I will clone this branch and try it out!

File: src/main/java/use_cases/add_edit_expenses_use_case/ExpenseID.java
Comment (YinLingithub) [reply src/main/java/use_cases/add_edit_expenses_use_case/ExpenseID.java:55]: She does. I've asked the same thing in her other feature. I'll resolve this to save time.

File: src/main/java/use_cases/add_edit_expenses_use_case/ExpenseUCI.java
Comment (katarinaavucic) [inline src/main/java/use_cases/add_edit_expenses_use_case/ExpenseUCI.java:149]: Nice fix!

**Review by katarinaavucic — DISMISSED** [review]
Approving so that we can merge to main and fix additional bugs later.

**Review by aricasas — DISMISSED** [review]
Approving current state so code is available on main

**Review by YinLingithub — DISMISSED** [review]
Approving so I could implement code for another pull request.

**Review by aricasas — APPROVED** [review]
Re-approved after merge conflicts

**Review by katarinaavucic — APPROVED** [review]
Approved to allow for integration with rest of code while we fix bugs.

**Review by YinLingithub — APPROVED** [review]
Reapprove after merge in local.

**Pull request merged**
